include/mutex.h

```diff
@@ -0,0 +1,37 @@
+#ifndef _MUTEX_
+#define _MUTEX_
+
+#include <atomic>
+
+//! Hybrid SpinMutex class
+//! \brief SpinMutex class that offers a fast locking and unlocking mechanism
+class SpinMutex {
+
+ public:
+  //! Attempt locking
+  bool try_lock() { return !lock_.test_and_set(std::memory_order_acquire); }
+
+  //! Call lock
+  void lock() {
+    std::size_t spin_count{0};
+    while (!try_lock()) {
+      ++spin_count;
+
+      if (spin_count < spin_pred_ * 2) continue;
+
+      std::this_thread::sleep_for(std::chrono::nanoseconds(1));
+    }
+
+    spin_pred_ += (spin_count - spin_pred_) / 8;
```

Comment:
I guess this is a predictor from measured value from the literature? I am just not sure where number 8 is from.